[Attributor] Use known non-flat AS before `getAssumedAddrSpace`

shiltian opened the pull request:
If the underlying object already has a non-flat address space, we simply use
that before calling `getAssumedAddrSpace`.

Partially fixes SWDEV-536263.

shiltian commented:
* **#143221** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/143221?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/143221?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

arsenm commented on llvm/lib/Transforms/IPO/AttributorAttributes.cpp:
```diff
@@ -12592,19 +12592,38 @@ struct AAAddressSpaceImpl : public AAAddressSpace {
 
     auto CheckAddressSpace = [&](Value &Obj) {
-      if (isa<UndefValue>(&Obj))
+      // Ignore undef and poison.
+      if (isa<UndefValue>(&Obj) || isa<PoisonValue>(&Obj))
         return true;
```
`isa<UndefValue>` already covers undef and poison. However, I think in this context it's unsafe to handle undef. You shouldn't touch this in this PR

shiltian committed:
[Attributor] Use known non-flat AS before `getAssumedAddrSpace`

If the underlying object already has a non-flat address space, we simply use
that before calling `getAssumedAddrSpace`.

shiltian commented:
The test failures have been fixed in 64bd4d91efb2de5380c9e61ff75c970293a11886.
